src/main/java/org/elasticsearch/index/translog/fs/FsChannelReader.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.translog.fs;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A base class for all classes that allows reading ops from translog files
+ */
+public abstract class FsChannelReader implements Closeable {
+
+    public static final int UNKNOWN = -1;
+
+    protected final long id;
+    protected final ChannelReference channelReference;
+    protected final FileChannel channel;
+    protected final AtomicBoolean closed = new AtomicBoolean(false);
+
+    public FsChannelReader(long id, ChannelReference channelReference) {
+        this.id = id;
+        this.channelReference = channelReference;
+        this.channel = channelReference.channel();
+        assert channelReference.assertAttach(this);
+    }
+
+    public long translogId() {
+        return this.id;
+    }
+
+    abstract public long sizeInBytes();
+
+    /** the position the first operation is written at */
+    public long firstPosition() {
+        return channelReference.stream().headerSize();
+    }
+
+
+    abstract public int estimatedTotalOperations();
+
+    public Translog.Operation readOp(Translog.Location location) throws IOException {
+        assert location.translogId == id : "read location's translog id [" + location.translogId + "] is not [" + id + "]";
+        ByteBuffer buffer = ByteBuffer.allocate(location.size);
+        return readOp(buffer, location.translogLocation, location.size);
+    }
+
+    /**
+     * reads an operation from the given position and changes to point at the start of the next
+     * operation
+     */
+    public Translog.Operation readOpAndAdvancePosition(ByteBuffer reusableBuffer, AtomicLong position) throws IOException {
+        final long pos = position.get();
```

Comment:
I have two return values from this method - a new location and the op. I can a return a tuple but that made me cringe (double object reference). All of this is just a question of style - if you prefer a tuple return, I'll change it.